internal/transports/p2p/server.go

```diff
@@ -29,99 +43,271 @@ func NewServer(
+			Msgf("peer banned, reason: %v", err)
+
+		s.addresses.BanAddr(&wire.NetAddress{IP: netAddr.IP, Port: uint16(netAddr.Port)})
 		return err
 	}
 
-	inbound := true
-	conn, err := listener.Accept()
+	peer, err := s.connectPeer(conn, false)
 	if err != nil {
-		s.log.Error().Msgf("error accepting connection, reason: %v", err)
+		s.log.Info().Str("peer", netAddr.String()).
+			Bool("inbound", false).
+			Msgf("peer banned, reason: %v", err)
+
+		s.addresses.BanAddr(&wire.NetAddress{IP: netAddr.IP, Port: uint16(netAddr.Port)})
 		return err
 	}
 
-	return s.connectPeer(conn, inbound)
+	_ = s.outboundPeers.AddPeer(peer) // don't need to check error here
+	s.addresses.UpsertPeerAddr(peer)
+	return nil
 }
 
-func (s *server) connectPeer(conn net.Conn, inbound bool) error {
-	peer, err := peer.NewPeer(conn, inbound, s.config, s.chainParams, s.headersService, s.chainService, s.log)
+func (s *server) connectPeer(conn net.Conn, inbound bool) (*peer.Peer, error) {
+	peer := peer.NewPeer(conn, inbound, s.config, s.chainParams, s.headersService, s.chainService, s.log, s)
+
+	err := peer.Connect()
 	if err != nil {
-		return err
+		peer.Disconnect()
+		s.log.Error().Str("peer", peer.String()).
+			Bool("inbound", inbound).
+			Msgf("error connecting with peer, reason: %v", err)
+		return nil, err
 	}
 
-	s.peers = append(s.peers, peer)
+	peer.SendGetAddrInfo()
 
-	err = peer.Connect()
+	if !inbound {
+		err = peer.StartHeadersSync()
+		if err != nil {
+			peer.Disconnect()
+			return nil, err
+		}
+	}
+
+	return peer, nil
+}
+
+func (s *server) observeOutboundPeers() {
+	sleeDuration := 1 * time.Minute
+	time.Sleep(sleeDuration)
+
+	for {
+		select {
+		case <-s.ctx.Done(): // Exit if context was canceled
+			s.log.Info().Msg("[observeOutboundPeers] context canceled -> exit")
+			return
+		default:
+			s.ctxWg.Add(1) // Wait on shutdown
+
+			freeSlots := s.outboundPeers.Space()
+			if freeSlots == 0 {
+				s.log.Debug().Msg("[observeOutboundPeers] nothing to do")
+				s.noWaitingSleep(sleeDuration)
+				continue
+			}
+
+			s.log.Info().Msgf("try connect with a new peer. Free slots: %d", freeSlots)
+			s.connectToRandomAddr() // Connect one-by-one to gracefully handle shutdown
+
+			s.ctxWg.Done()
+		}
+	}
+}
+
+func (s *server) connectToRandomAddr() {
+	const tries = 20
+	addr := s.addresses.GetRndUnusedAddr(tries)
+	if addr == nil {
+		s.log.Warn().Msgf("[observeOutboundPeers] coudnt find random unused/unbanned peer address with %d tries", tries)
+		return
+	}
+
+	_ = s.connectToAddr(addr.IP, addr.Port)
+}
+
+func (s *server) observeInboundPeers() {
+	sleeDuration := 1 * time.Minute
+	time.Sleep(sleeDuration)
+
+	for {
+		select {
+		case <-s.ctx.Done(): // Exit if context was canceled
+			s.log.Info().Msg("[observeInboundPeers] context canceled -> exit")
+			return
+		default:
+			s.ctxWg.Add(1) // Wait on shutdown
+
+			if s.inboundPeers.Space() == 0 {
+				s.log.Debug().Msg("[observeInboundPeers] nothing to do")
+				s.noWaitingSleep(sleeDuration)
```

Comment:
We don't need to sleep here, because the operation `s.listener.Accept` in the `s.waitForIncomingConnection()` function is blocking.

Reply:
Waiting (at least for now) is necessary here because we want to wait for some time before the next check to see if we have a free slot for an incoming connection.

Reply:
implementation changed